capabilities/controller/pkg/capabilities/core/capability_controller.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package core
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	corev1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/core/v1alpha1"
+	"github.com/vmware-tanzu/tanzu-framework/capabilities/client/pkg/discovery"
+	"github.com/vmware-tanzu/tanzu-framework/capabilities/controller/pkg/config"
+	"github.com/vmware-tanzu/tanzu-framework/capabilities/controller/pkg/constants"
+)
+
+// CapabilityReconciler reconciles a Capability object.
+type CapabilityReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+	Host   string
+}
+
+//+kubebuilder:rbac:groups=run.tanzu.vmware.com,resources=capabilities,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=run.tanzu.vmware.com,resources=capabilities/status,verbs=get;update;patch
+
+// Reconcile reconciles a Capability spec by executing specified queries.
+func (r *CapabilityReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctxCancel, cancel := context.WithTimeout(ctx, constants.ContextTimeout)
+	defer cancel()
+
+	log := r.Log.WithValues("capability", req.NamespacedName)
+	log.Info("Starting reconcile")
+
+	capability := &corev1alpha1.Capability{}
+	if err := r.Get(ctxCancel, req.NamespacedName, capability); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	var serviceAccountName, namespaceName string
+	// use the default service account when the serviceAccountName is not provided as part of the spec
+	if len(capability.Spec.ServiceAccountName) > 0 {
+		serviceAccountName = capability.Spec.ServiceAccountName
+		namespaceName = req.Namespace
+	} else {
+		serviceAccountName = constants.ServiceAccountWithDefaultPermissions
+		namespaceName = constants.CapabilitiesControllerNamespace
+	}
+	config, err := config.GetConfigForServiceAccount(ctx, r.Client, namespaceName, serviceAccountName, r.Host)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to get config for ClusterQueryClient creation: %w", err)
```

Comment:
nit: I think error message would be enough to be like,
```fmt.Errorf("get config for ClusterQueryClient creation: %w", err)```

My thinking is, as it tries to get a config, an error will appear why it couldn't be got. Saying "unable to" seems redundant. The following error message indicates the config was unable to be gotten.

Same goes for line 60. Perhaps, "unable to create..." could be simple, "create ClusterQueryClient:".
But this is a matter of preference. It's optional.

Reply:
I prefer using `failed to` or `unable to` phrases in error messages as it is slightly more clear and doesn't feel like an incomplete sentence.